File: app/src/main/java/com/shevelev/comics_viewer/activities/main/bookshelf/BookshelfComicsReader.java
package com.shevelev.comics_viewer.activities.main.bookshelf;

import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.util.Log;

import com.shevelev.comics_viewer.R;
import com.shevelev.comics_viewer.activities.main.comics_filters.BookcaseComics;
import com.shevelev.comics_viewer.activities.main.comics_filters.IComicsFilter;
import com.shevelev.comics_viewer.comics_workers.CoverCreator;
import com.shevelev.comics_viewer.comics_workers.IPreviewCreator;
import com.shevelev.comics_viewer.comics_workers.PreviewCreator;
import com.shevelev.comics_viewer.common.func_interfaces.IActionOneArgs;
import com.shevelev.comics_viewer.common.func_interfaces.IActionZeroArgs;
import com.shevelev.comics_viewer.common.helpers.BitmapsHelper;
import com.shevelev.comics_viewer.common.helpers.CollectionsHelper;
import com.shevelev.comics_viewer.common.helpers.files.AppPrivateFilesHelper;
import com.shevelev.comics_viewer.common.structs.Size;

import java.util.List;

/**
 * Read comics for bookshelf
 */
public class BookshelfComicsReader  extends AsyncTask<Void, Void, Void>
{
    private List<BookshelfComicsInfo> readComics;

    private IActionZeroArgs beforeExecute;
    private IActionOneArgs<List<BookshelfComicsInfo>> afterExecute;

    private IComicsFilter comicsFilter;     // source of comics

    private Bitmap privateCover;

    private final IPreviewCreator previewCreator;

    public BookshelfComicsReader(
            IComicsFilter comicsFilter,
            IActionZeroArgs beforeExecute,
            IActionOneArgs<List<BookshelfComicsInfo>> afterExecute,
            Size clientSize)
    {
        this.beforeExecute = beforeExecute;
        this.afterExecute = afterExecute;
        this.comicsFilter = comicsFilter;

        previewCreator = new PreviewCreator(clientSize);
    }

    @Override
    protected Void doInBackground(Void... params)
    {
        try
        {
            List<BookcaseComics> comics=comicsFilter.getComics();

            if(comics!=null)
                readComics=CollectionsHelper.transform(comics, item ->
                        new BookshelfComicsInfo(
                                item.id,
                                item.displayName,
                                getCover(item),
                                item.isNeedShowPrivateCover));

            else
                readComics = null;
        }
        catch (Exception ex)
        {
            Log.e("CV", "exception", ex);
        }

        return null;
    }

    private Bitmap getCover(BookcaseComics comics)
    {
        if(comics.isNeedShowPrivateCover)
        {
            if(privateCover==null)
                privateCover = CoverCreator.create(BitmapsHelper.loadFromRaw(R.drawable.img_private_comics_cover), previewCreator);

            return privateCover;
        }

        return BitmapsHelper.loadFromFile(AppPrivateFilesHelper.getFullName(comics.coverFilename));
    }

    @Override
    protected void onPreExecute()
    {
        super.onPreExecute();

        beforeExecute.process();
    }

    @Override
    protected void onPostExecute(Void result)
    {
        super.onPostExecute(result);

        afterExecute.process(readComics);
    }
}

File: app/src/main/java/com/shevelev/comics_viewer/activities/main/bookshelf/BookshelfComicsReader.kt
package com.shevelev.comics_viewer.activities.main.bookshelf

import android.graphics.Bitmap
import android.os.AsyncTask
import android.util.Log
import com.shevelev.comics_viewer.R
import com.shevelev.comics_viewer.activities.main.comics_filters.BookcaseComics
import com.shevelev.comics_viewer.activities.main.comics_filters.IComicsFilter
import com.shevelev.comics_viewer.comics_workers.CoverCreator
import com.shevelev.comics_viewer.comics_workers.IPreviewCreator
import com.shevelev.comics_viewer.comics_workers.PreviewCreator
import com.shevelev.comics_viewer.common.func_interfaces.IActionOneArgs
import com.shevelev.comics_viewer.common.func_interfaces.IActionZeroArgs
import com.shevelev.comics_viewer.common.helpers.BitmapsHelper
import com.shevelev.comics_viewer.common.helpers.CollectionsHelper
import com.shevelev.comics_viewer.common.helpers.files.AppPrivateFilesHelper
import com.shevelev.comics_viewer.common.structs.Size

/**
 * Read comics for bookshelf
 */
class BookshelfComicsReader(
    // source of comics
    private val comicsFilter: IComicsFilter,
    private val beforeExecute: IActionZeroArgs,
    private val afterExecute: IActionOneArgs<List<BookshelfComicsInfo>?>,
    clientSize: Size) : AsyncTask<Void?, Void?, Void?>() {

    private var readComics: List<BookshelfComicsInfo>? = null
    private var privateCover: Bitmap? = null
    private val previewCreator: IPreviewCreator = PreviewCreator(clientSize)

    protected override fun doInBackground(vararg params: Void?): Void? {
        try {
            val comics = comicsFilter.comics
            readComics = if (comics != null) CollectionsHelper.transform(comics) { item: BookcaseComics ->
                BookshelfComicsInfo(
                    item.id,
                    item.displayName!!,
                    getCover(item)!!,
                    item.isNeedShowPrivateCover)
            } else null
        } catch (ex: Exception) {
            Log.e("CV", "exception", ex)
        }
        return null
    }

    private fun getCover(comics: BookcaseComics): Bitmap? {
        if (comics.isNeedShowPrivateCover) {
            if (privateCover == null) privateCover = CoverCreator.create(BitmapsHelper.loadFromRaw(R.drawable.img_private_comics_cover), previewCreator)
            return privateCover
        }
        return BitmapsHelper.loadFromFile(AppPrivateFilesHelper.getFullName(comics.coverFilename))
    }

    override fun onPreExecute() {
        super.onPreExecute()
        beforeExecute.process()
    }

    override fun onPostExecute(result: Void?) {
        super.onPostExecute(result)
        afterExecute.process(readComics)
    }
}